metrics.go

```diff
@@ -80,7 +73,48 @@ func newMetrics() Metrics {
 	}
 }
 
-func newErrorsTotalMap() map[string]*uint64 {
-	m := make(map[string]*uint64, 20)
-	return m
+// ErrorsTotalMap stores the counter for the different http errors.
+type ErrorsTotalMap struct {
+	sync.RWMutex
+	m map[HTTPError]*uint64
+}
+
+func newErrorsTotalMap() ErrorsTotalMap {
+	m := make(map[HTTPError]*uint64, 20)
+	return ErrorsTotalMap{
+		m: m,
+	}
+}
+
+// incErrorsTotal increases the counter for this error atomically by one.
+func (e *ErrorsTotalMap) incError(err HTTPError) {
+	// The goal is to have a valid ptr to the counter for this err
+	e.RLock()
+	ptr, ok := e.m[err]
+	e.RUnlock()
+	if !ok {
+		// The ptr does not seem to exist for this err
+		// Hence we create it (using a write lock)
+		e.Lock()
+		// We ensure that the ptr wasn't created in the meantime
+		if ptr, ok = e.m[err]; !ok {
+			ptr = new(uint64)
+			*ptr = 0
```

Comment:
Initializing a pointer to its default value after `new()` is not necessary. It will be automatically done, see the Go spec: https://golang.org/ref/spec#Allocation

> The variable is initialized as described in the section on initial values.